support at::view and at::reshape for quantized tensor

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23121 [quant] Verify flatten works for quantized Tensor
* **#23046 support at::view and at::reshape for quantized tensor**
* #23045 Move THTensor_compute_stride to ATen

att

Differential Revision: [D16368897](https://our.internmc.facebook.com/intern/diff/D16368897/)

jerryzh168 committed:
support at::view and at::reshape for quantized tensor

att

Differential Revision: [D16368897](https://our.internmc.facebook.com/intern/diff/D16368897/)

dzhulgakov commented on aten/src/ATen/native/quantized/TensorShape.cpp:
```diff
@@ -0,0 +1,29 @@
+namespace at {
+namespace native {
+
+Tensor quantized_view(const Tensor& self, IntArrayRef size) {
```
btw, is there anything qtensor specific here? or is it just that you refactored it out of TH? if so - maybe call it generically and also route this function for CPU/GPU to ATen instead of _th_view? (just in spirit of improving codebase)

jerryzh168 replied:
not really, I can remove _th_view

jerryzh168 replied:
then should I put this somewhere else?

dzhulgakov commented on test/test_quantized_tensor.py:
```diff
@@ -202,5 +198,48 @@ def test_qtensor_clone(self):
+        b = a.transpose(1, 2)  # swaps 2nd and 3rd dimension
+        c = a.view(1, 3, 2, 4)  # does not change tensor layout
+        self.assertEqual(b.size(), c.size())
+        self.assertNotEqual(b.int_repr(), c.int_repr())
```
I'd also check that quantizer is the same

dzhulgakov reviewed (approved):
Approving, but it'd be nice to do a cleanup

jerryzh168 committed:
Update on "support at::view and at::reshape for quantized tensor"


att

Differential Revision: [D16368897](https://our.internmc.facebook.com/intern/diff/D16368897/)

li-roy reviewed (approved):
Nice, thanks for porting view!

jerryzh168 committed:
Update on "support at::view and at::reshape for quantized tensor"


att

Differential Revision: [D16368897](https://our.internmc.facebook.com/intern/diff/D16368897/)

jerryzh168 committed:
Update on "support at::view and at::reshape for quantized tensor"


att

Differential Revision: [D16368897](https://our.internmc.facebook.com/intern/diff/D16368897/)

gchanan commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -896,6 +896,24 @@ Tensor numpy_T(const Tensor &self) {
+    "not compatible with input tensor's size and stride (at least one dimension"
+    " spans across two contiguous subspaces). Use .reshape(...) instead.");
+  auto stride_value = *stride;
+  auto self_ = self.clone();
```
wait, we are cloning in view?!?!?!

gchanan replied:
can we _please_ do our best to have a single implementation in TH and ATen?  You can call ATen from TH, so there's usually no blocker to having a single implementation (use `THTensor_wrap` -- the only issue you might run into is extra refcounting in an inner loop, but that's pretty unlikely for `view`).

But if you had a single implementation (i.e. moved the existing one) it would be obvious that an extra `clone` call was added.

jerryzh168 replied:
you mean refactor TH code to call ATen function?

dzhulgakov replied:
Sorry, my bad - I missed the clone() in review.

Re sharing - my comment was about exactly that - redirecting to the same implementation, i.e. either using _th_view everywhere or (better) removing _th_view all together and just calling ATen function in THTensor_(newView)

jerryzh168 replied:
I checked codebase, looks like there is no more reference to _th_view in aten, so I'm using at::view in TH code and removing _th_view code in my fix diff.
